src/lightning/data/streaming/combined.py

```diff
@@ -0,0 +1,93 @@
+# Copyright The Lightning AI team.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+from typing import Any, Dict, Iterator, List, Optional, Sequence
+
+from torch.utils.data import IterableDataset
+
+from lightning.data.streaming.dataset import StreamingDataset
+
+
+class CombinedStreamingDataset(IterableDataset):
+    """The `CombinedStreamingDataset` enables to stream data from multiple StreamingDataset with the sampling ratio of
+    your choices.
+
+    Addtionally, the `CombinedStreamingDataset` keeps track of the number of
+    samples fetched to enable resumability of the datasets.
+
+    """
+
+    def __init__(
+        self, datasets: List[StreamingDataset], seed: int = 42, weights: Optional[Sequence[float]] = None
+    ) -> None:
+        self._seed = seed
+        self._datasets = datasets
+        self._weights = weights
+        num_datasets = len(datasets)
+
+        # Inversely weighted based on length
+        if weights is None:
+            self._weights = [1 / float(num_datasets)] * num_datasets
+        else:
+            self._weights = [w / sum(weights) for w in weights]
+
+        self._iterator: Optional[_CombinedDatasetIterator] = None
+
+    def __iter__(self) -> Iterator[Any]:
+        assert self._weights
+        self._iterator = _CombinedDatasetIterator(self._datasets, self._seed, self._weights)
+        return self._iterator
+
+    def state_dict(self) -> Dict[str, Any]:
+        if self._iterator is None:
+            return {}
+        return self._iterator.state_dict()
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        if len(state_dict) != len(self._datasets):
+            raise RuntimeError(
+                f"The provided state {state_dict} doesn't match the current number of datasets: {self._datasets}."
+            )
+
+        for dataset_idx, dataset in enumerate(self._datasets):
+            if str(dataset_idx) not in state_dict:
+                raise RuntimeError(f"The provided state {state_dict} doesn't contain the index {dataset_idx}.")
```

Comment:
I don't think we should print the entire state dict here, this will make it really hard to read the error message otherwise.
```suggestion
                raise RuntimeError(f"The provided state doesn't contain the index {dataset_idx}.")
```

Reply:
Sure.